adafruit_as7341.py

```diff
@@ -43,67 +43,78 @@
+_AS7341_CH0_DATA_L: int = const(0x95)  # ADC Channel 0 Data
+_AS7341_CH0_DATA_H: int = const(0x96)  # ADC Channel 0 Data
+_AS7341_CH1_DATA_L: int = const(0x97)  # ADC Channel 1 Data
+_AS7341_CH1_DATA_H: int = const(0x98)  # ADC Channel 1 Data
+_AS7341_CH2_DATA_L: int = const(0x99)  # ADC Channel 2 Data
+_AS7341_CH2_DATA_H: int = const(0x9A)  # ADC Channel 2 Data
+_AS7341_CH3_DATA_L: int = const(0x9B)  # ADC Channel 3 Data
+_AS7341_CH3_DATA_H: int = const(0x9C)  # ADC Channel 3 Data
+_AS7341_CH4_DATA_L: int = const(0x9D)  # ADC Channel 4 Data
+_AS7341_CH4_DATA_H: int = const(0x9E)  # ADC Channel 4 Data
+_AS7341_CH5_DATA_L: int = const(0x9F)  # ADC Channel 5 Data
+_AS7341_CH5_DATA_H: int = const(0xA0)  # ADC Channel 5 Data
+_AS7341_STATUS2: int = const(0xA3)  # Measurement status flags; saturation, validity
+_AS7341_STATUS3: int = const(0xA4)  # Spectral interrupt source, high or low threshold
+_AS7341_CFG0: int = const(
     0xA9
 )  # Sets Low power mode, Register bank, and Trigger lengthening
-_AS7341_CFG1 = const(0xAA)  # Controls ADC Gain
-_AS7341_CFG6 = const(0xAF)  # Used to configure Smux
-_AS7341_CFG9 = const(0xB2)  # flicker detect and SMUX command system ints
-_AS7341_CFG12 = const(0xB5)  # ADC channel for interrupts, persistence and auto-gain
+_AS7341_CFG1: int = const(0xAA)  # Controls ADC Gain
+_AS7341_CFG6: int = const(0xAF)  # Used to configure Smux
+_AS7341_CFG9: int = const(0xB2)  # flicker detect and SMUX command system ints
+_AS7341_CFG12: int = const(
+    0xB5
+)  # ADC channel for interrupts, persistence and auto-gain
 _AS7341_PERS = const(
     0xBD
 )  # number of measurements outside thresholds to trigger an interrupt
 _AS7341_GPIO2 = const(
     0xBE
 )  # GPIO Settings and status: polarity, direction, sets output, reads
-_AS7341_ASTEP_L = const(0xCA)  # Integration step size ow byte
-_AS7341_ASTEP_H = const(0xCB)  # Integration step size high byte
-_AS7341_FD_TIME1 = const(0xD8)  # Flicker detection integration time low byte
-_AS7341_FD_TIME2 = const(0xDA)  # Flicker detection gain and high nibble
-_AS7341_FD_STATUS = const(
+_AS7341_ASTEP_L: int = const(0xCA)  # Integration step size ow byte
+_AS7341_ASTEP_H: int = const(0xCB)  # Integration step size high byte
+_AS7341_FD_TIME1: int = const(0xD8)  # Flicker detection integration time low byte
+_AS7341_FD_TIME2: int = const(0xDA)  # Flicker detection gain and high nibble
+_AS7341_FD_STATUS: int = const(
     0xDB
 )  # Flicker detection status; measurement valid, saturation, flicker
-_AS7341_INTENAB = const(0xF9)  # Enables individual interrupt types
-_AS7341_CONTROL = const(0xFA)  # Auto-zero, fifo clear, clear SAI active
-_AS7341_FD_CFG0 = const(0xD7)  # Enables FIFO for flicker detection
+_AS7341_INTENAB: int = const(0xF9)  # Enables individual interrupt types
+_AS7341_CONTROL: int = const(0xFA)  # Auto-zero, fifo clear, clear SAI active
+_AS7341_FD_CFG0: int = const(0xD7)  # Enables FIFO for flicker detection
 
 
-def _low_bank(func):
+def _low_bank(func) -> any:
```

Comment:
Oh I hate typing decorators :'( I looked up how to do this: https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#decorators

So it seems like in that typing block at the top of the file, after the `typing` import, you should declare a type variable based on `Callable` (which is typing for methods and functions):

`TCallable = TypeVar('TCallable', bound=Callable[..., Any])`

Then, here, you can type it as follows:

`def _low_bank(func: TCallable) -> TCallable`

That shoooooould work but apologies in advance if it doesn't!  We'll get it to work if that doesn't.

Reply:
It seems like there ought to be a tidier way to do that, but apparently not. This got me started on the road to the correct solution (or at least, the one which made the linter stop complaining). 

I had to define `TCallable` as you suggested, but then I had to declare the inner `_decorator` function as returning a `TCallable`, and I had to declare that the `_low_bank` function took a TCallable argument and returned a `Callable[[Any], TCallable]`. I confess I'm not sure I understand decorators well enough to understand why that works, but it seems to satisfy the linter.

Reply:
Hmm... Interesting that that didn't work.  I don't think the inner `_decorator` should return a function so it should be `Callable` or `TCallable`, I think.  So this didn't work? :

```
def _low_bank(func: TCallable) -> TCallable:
    # pylint:disable=protected-access
    def _decorator(self, *args, **kwargs)
        self._low_bank_active = True
        retval = func(self, *args, **kwargs)
        self._low_bank_active = False
        return retval

    return _decorator
```

I was doing some thinking, what about this:

```
def _low_bank(func: TCallable) -> Callable[[Any], Any]:
    # pylint:disable=protected-access
    def _decorator(self, *args, **kwargs) -> Any:
        self._low_bank_active = True
        retval = func(self, *args, **kwargs)
        self._low_bank_active = False
        return retval

    return _decorator
```

I believe `_decorator` should return anything so `Any`, but `_low_bank`, as the decorator should take and return methods, so it should be something like the above.  It may also be work not typing this, too, since it looks like the decorator is mostly used for internal functionality and isn't outward facing...

Reply:
Your first example didn't work, in the sense that the linter gave me a weird error about expecting a `Callable[[Any, VarArg[Any], KwArg[Any], Any], Any]` but seeing a `TCallable` instead (I think; I'm not in front of my computer right now so I'm going from memory). I admit I'm not sure my understanding of Callable things is deep enough to fully understand what the linter is complaining about. 

I'd also be okay with just removing the type hints on that whole chunk of code since `_low_bank` is internal, though this feels a bit like letting the linter win the argument. :) I don't have a strong preference here one way or the other, so I'm open for input.

Reply:
@FoamyGuy, any thoughts?  I've run into issues with typing my own decorators in the past, so I'm partial to just not typing.  Especially because it doesn't seem to be used to change what the functions it affects return, so them being typed correctly should be sufficient.  Would love to type it correctly, but also happy to let the linter win the battle so we can win the war 😂

Reply:
I also don't have too much knowledge or experience around decorators. I would be fine with skipping it or using something general like `Any` for the time being. Right now I think we are still working on "broad strokes" of the typing effort. Totally fine if we don't have 100% of everything completed in the perfect manner. The types can always be expanded or improved upon in future PRs if opportunity arises.

Reply:
Sounds good. Thank you both! I'll change the type hint to `Any` and commit.

Reply:
I don't think `_decorator` should return a `Callable` or `TypeVar` of `Callable` since it "represents" the newly decorated function, so it should return something more like `Any` since we don't know what the pre-decorated function `func` does.  So I agree with you @FoamyGuy on your suggestions to use `Any`s or not type.

Either way, its a cool use of decorators though!